File: reference/python/aicp_ref_capneg_v02/profile_composition.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from aicp_ref.hashing import object_hash


ROOT = Path(__file__).resolve().parents[3]
DEFAULT_RULES_PATH = ROOT / "registry/aicp_profile_composition_rules.json"
COMPOSITION_VERSION = "aicp.profile_composition.v1"
COMPOSITION_HASH_DOMAIN = "capneg.profile_composition"


def canonical_profile_ref_key(value: Any) -> tuple[str, str]:
    if not isinstance(value, dict):
        return "", ""
    return str(value.get("profile_id", "")), str(value.get("profile_version", ""))


def _profile_label(value: Any) -> str:
    profile_id, profile_version = canonical_profile_ref_key(value)
    return f"{profile_id}@{profile_version}"


def _error(code: str, detail: str) -> dict[str, str]:
    return {"code": code, "detail": detail}


def load_composition_rules(path: Path = DEFAULT_RULES_PATH) -> dict[str, Any]:
    payload = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(payload, dict):
        raise ValueError("profile-composition rules registry must be an object")
    return payload


def resolve_profile_composition(
    composition: Any,
    rules: dict[str, Any] | None = None,
) -> dict[str, Any]:
    rules = rules or load_composition_rules()
    errors: list[dict[str, str]] = []

    if not isinstance(composition, dict):
        composition = {}
        errors.append(
            _error(
                "PROFILE_COMPOSITION_SHAPE_INVALID",
                "profile composition must be an object",
            )
        )
    if composition.get("composition_version") != COMPOSITION_VERSION:
        errors.append(
            _error(
                "PROFILE_COMPOSITION_VERSION_UNSUPPORTED",
                f"composition_version must equal {COMPOSITION_VERSION}",
            )
        )

    raw_profiles = composition.get("profiles")
    if not isinstance(raw_profiles, list):
        raw_profiles = []
        errors.append(
            _error(
                "PROFILE_COMPOSITION_SHAPE_INVALID",
                "profiles must be an array",
            )
        )
    if not raw_profiles:
        errors.append(
            _error(
                "PROFILE_COMPOSITION_EMPTY",
                "profile composition must contain at least one profile",
            )
        )
    maximum = int(rules.get("maximum_profiles", 16))
    if len(raw_profiles) > maximum:
        errors.append(
            _error(
                "PROFILE_COMPOSITION_LIMIT_EXCEEDED",
                f"profile composition exceeds maximum_profiles={maximum}",
            )
        )

    valid_shape = all(
        isinstance(profile, dict)
        and set(profile) == {"profile_id", "profile_version"}
        and isinstance(profile.get("profile_id"), str)
        and bool(profile["profile_id"])
        and isinstance(profile.get("profile_version"), str)
        and bool(profile["profile_version"])
        for profile in raw_profiles
    )
    if not valid_shape:
        errors.append(
            _error(
                "PROFILE_REF_INVALID",
                "every profile must contain only non-empty profile_id and profile_version",
            )
        )

    canonical_profiles = sorted(raw_profiles, key=canonical_profile_ref_key)
    canonical_composition = {
        "composition_version": COMPOSITION_VERSION,
        "profiles": canonical_profiles,
    }
    if raw_profiles != canonical_profiles:
        errors.append(
            _error(
                "PROFILE_ORDER_NON_CANONICAL",
                "profiles must be sorted by exact profile_id then profile_version",
            )
        )

    keys = [canonical_profile_ref_key(profile) for profile in raw_profiles]
    duplicate_keys = sorted({key for key in keys if keys.count(key) > 1})
    for key in duplicate_keys:
        errors.append(
            _error(
                "PROFILE_DUPLICATE",
                f"duplicate profile reference {key[0]}@{key[1]}",
            )
        )

    versions_by_id: dict[str, set[str]] = {}
    for profile_id, profile_version in keys:
        versions_by_id.setdefault(profile_id, set()).add(profile_version)
    for profile_id in sorted(versions_by_id):
        versions = sorted(versions_by_id[profile_id])
        if len(versions) > 1:
            errors.append(
                _error(
                    "PROFILE_FAMILY_VERSION_CONFLICT",
                    f"{profile_id} selects multiple exact versions: {versions}",
                )
            )

    records = {
        canonical_profile_ref_key(record["profile"]): record
        for record in rules.get("profiles", [])
        if isinstance(record, dict) and isinstance(record.get("profile"), dict)
    }
    selected_records: list[dict[str, Any]] = []
    for key in sorted(set(keys)):
        record = records.get(key)
        if record is None:
            errors.append(
                _error(
                    "PROFILE_UNKNOWN",
                    f"unknown exact profile {key[0]}@{key[1]}",
                )
            )
        else:
            selected_records.append(record)

    core_paths = sorted(
        {
            str(record.get("core_suite", {}).get("path", ""))
            for record in selected_records
            if isinstance(record.get("core_suite"), dict)
        }
    )
    if len(core_paths) > 1:
        errors.append(
            _error(
                "PROFILE_CORE_VERSION_CONFLICT",
                f"selected profiles resolve to multiple Core suites: {core_paths}",
            )
        )
    for record in selected_records:
        if record.get("negotiable_by_capneg_v0_2") is not True:
            reason = record.get("capneg_v0_2_unsupported_reason", {})
            errors.append(
                _error(
                    str(reason.get("reason_code", "CAPNEG_CORE_FAMILY_UNSUPPORTED")),
                    str(
                        reason.get(
                            "detail",
                            f"{_profile_label(record['profile'])} is not negotiable by CAPNEG v0.2",
                        )
                    ),
                )
            )

    selected_keys = set(keys)
    subset_relations = rules.get("rules", {}).get(
        "strict_suite_subset_relations", []
    )
    for relation in subset_relations:
        redundant = canonical_profile_ref_key(relation.get("redundant_profile"))
        covering = canonical_profile_ref_key(relation.get("covering_profile"))
        if redundant in selected_keys and covering in selected_keys:
            errors.append(
                _error(
                    "PROFILE_COMPOSITION_REDUNDANT",
                    f"{redundant[0]}@{redundant[1]} is a strict required-suite "
                    f"subset of {covering[0]}@{covering[1]}",
                )
            )

    for group in rules.get("rules", {}).get("exclusive_groups", []):
        members = {
            canonical_profile_ref_key(member) for member in group.get("members", [])
        }
        selected_members = sorted(selected_keys & members)
        maximum_selected = int(group.get("max_selected", 1))
        if len(selected_members) > maximum_selected:
            errors.append(
                _error(
                    "PROFILE_COMPOSITION_EXCLUSIVE_CONFLICT",
                    f"exclusive group {group.get('group_id')} allows at most "
                    f"{maximum_selected}: {selected_members}",
                )
            )

    def union(field: str) -> list[str]:
        return sorted(
            {
                value
                for record in selected_records
                for value in record.get(field, [])
                if isinstance(value, str)
            }
        )

    component_marks = sorted(
        {
            record["compatibility_mark"]
            for record in selected_records
            if isinstance(record.get("compatibility_mark"), str)
        }
    )
    composition_hash: str | None = None
    if not errors:
        composition_hash = object_hash(
            COMPOSITION_HASH_DOMAIN, canonical_composition
        )

    return {
        "composition": canonical_composition,
        "composition_hash": composition_hash,
        "core_suite": (
            selected_records[0].get("core_suite")
            if len(core_paths) == 1 and selected_records
            else None
        ),
        "required_suites": union("required_suites"),
        "required_extensions": union("required_extensions"),
        "required_crypto_profiles": union("required_crypto_profiles"),
        "required_policy_categories": union("required_policy_categories"),
        "component_compatibility_marks": component_marks,
        "errors": errors,
    }

File: sdk/typescript/src/profile_composition.js
import { objectHash } from "./hashing.js";

export const COMPOSITION_VERSION = "aicp.profile_composition.v1";
export const COMPOSITION_HASH_DOMAIN = "capneg.profile_composition";

export function canonicalProfileRefKey(value) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return ["", ""];
  }
  return [String(value.profile_id ?? ""), String(value.profile_version ?? "")];
}

function compareProfileRefs(left, right) {
  const [leftId, leftVersion] = canonicalProfileRefKey(left);
  const [rightId, rightVersion] = canonicalProfileRefKey(right);
  return leftId.localeCompare(rightId) || leftVersion.localeCompare(rightVersion);
}

function profileKey(value) {
  return canonicalProfileRefKey(value).join("\0");
}

function error(code, detail) {
  return { code, detail };
}

function clone(value) {
  return JSON.parse(JSON.stringify(value));
}

export function resolveProfileComposition(compositionValue, rules) {
  let composition = compositionValue;
  const errors = [];
  if (composition === null || typeof composition !== "object" || Array.isArray(composition)) {
    composition = {};
    errors.push(error("PROFILE_COMPOSITION_SHAPE_INVALID", "profile composition must be an object"));
  }
  if (composition.composition_version !== COMPOSITION_VERSION) {
    errors.push(
      error(
        "PROFILE_COMPOSITION_VERSION_UNSUPPORTED",
        `composition_version must equal ${COMPOSITION_VERSION}`,
      ),
    );
  }
  let rawProfiles = composition.profiles;
  if (!Array.isArray(rawProfiles)) {
    rawProfiles = [];
    errors.push(error("PROFILE_COMPOSITION_SHAPE_INVALID", "profiles must be an array"));
  }
  if (rawProfiles.length === 0) {
    errors.push(
      error("PROFILE_COMPOSITION_EMPTY", "profile composition must contain at least one profile"),
    );
  }
  const maximum = Number(rules.maximum_profiles ?? 16);
  if (rawProfiles.length > maximum) {
    errors.push(
      error(
        "PROFILE_COMPOSITION_LIMIT_EXCEEDED",
        `profile composition exceeds maximum_profiles=${maximum}`,
      ),
    );
  }
  const validShape = rawProfiles.every(
    (profile) =>
      profile !== null &&
      typeof profile === "object" &&
      !Array.isArray(profile) &&
      Object.keys(profile).sort().join(",") === "profile_id,profile_version" &&
      typeof profile.profile_id === "string" &&
      profile.profile_id.length > 0 &&
      typeof profile.profile_version === "string" &&
      profile.profile_version.length > 0,
  );
  if (!validShape) {
    errors.push(
      error(
        "PROFILE_REF_INVALID",
        "every profile must contain only non-empty profile_id and profile_version",
      ),
    );
  }

  const canonicalProfiles = clone(rawProfiles).sort(compareProfileRefs);
  const canonicalComposition = {
    composition_version: COMPOSITION_VERSION,
    profiles: canonicalProfiles,
  };
  if (JSON.stringify(rawProfiles) !== JSON.stringify(canonicalProfiles)) {
    errors.push(
      error(
        "PROFILE_ORDER_NON_CANONICAL",
        "profiles must be sorted by exact profile_id then profile_version",
      ),
    );
  }
  const keys = rawProfiles.map(profileKey);
  for (const key of [...new Set(keys.filter((candidate) => keys.filter((value) => value === candidate).length > 1))].sort()) {
    const [profileId, profileVersion] = key.split("\0");
    errors.push(
      error("PROFILE_DUPLICATE", `duplicate profile reference ${profileId}@${profileVersion}`),
    );
  }
  const versionsById = new Map();
  for (const profile of rawProfiles) {
    const [profileId, profileVersion] = canonicalProfileRefKey(profile);
    if (!versionsById.has(profileId)) versionsById.set(profileId, new Set());
    versionsById.get(profileId).add(profileVersion);
  }
  for (const profileId of [...versionsById.keys()].sort()) {
    const versions = [...versionsById.get(profileId)].sort();
    if (versions.length > 1) {
      errors.push(
        error(
          "PROFILE_FAMILY_VERSION_CONFLICT",
          `${profileId} selects multiple exact versions: ${JSON.stringify(versions)}`,
        ),
      );
    }
  }

  const records = new Map(
    (rules.profiles ?? []).map((record) => [profileKey(record.profile), record]),
  );
  const selectedRecords = [];
  for (const key of [...new Set(keys)].sort()) {
    const record = records.get(key);
    if (record === undefined) {
      const [profileId, profileVersion] = key.split("\0");
      errors.push(error("PROFILE_UNKNOWN", `unknown exact profile ${profileId}@${profileVersion}`));
    } else {
      selectedRecords.push(record);
    }
  }
  const corePaths = [
    ...new Set(selectedRecords.map((record) => record.core_suite?.path ?? "")),
  ].sort();
  if (corePaths.length > 1) {
    errors.push(
      error(
        "PROFILE_CORE_VERSION_CONFLICT",
        `selected profiles resolve to multiple Core suites: ${JSON.stringify(corePaths)}`,
      ),
    );
  }
  for (const record of selectedRecords) {
    if (record.negotiable_by_capneg_v0_2 !== true) {
      const reason = record.capneg_v0_2_unsupported_reason ?? {};
      errors.push(
        error(
          String(reason.reason_code ?? "CAPNEG_CORE_FAMILY_UNSUPPORTED"),
          String(reason.detail ?? "profile is not negotiable by CAPNEG v0.2"),
        ),
      );
    }
  }
  const selectedKeys = new Set(keys);
  for (const relation of rules.rules?.strict_suite_subset_relations ?? []) {
    const redundant = profileKey(relation.redundant_profile);
    const covering = profileKey(relation.covering_profile);
    if (selectedKeys.has(redundant) && selectedKeys.has(covering)) {
      const [redundantId, redundantVersion] = redundant.split("\0");
      const [coveringId, coveringVersion] = covering.split("\0");
      errors.push(
        error(
          "PROFILE_COMPOSITION_REDUNDANT",
          `${redundantId}@${redundantVersion} is a strict required-suite subset of ${coveringId}@${coveringVersion}`,
        ),
      );
    }
  }
  for (const group of rules.rules?.exclusive_groups ?? []) {
    const members = new Set((group.members ?? []).map(profileKey));
    const selected = [...selectedKeys].filter((key) => members.has(key)).sort();
    const maxSelected = Number(group.max_selected ?? 1);
    if (selected.length > maxSelected) {
      errors.push(
        error(
          "PROFILE_COMPOSITION_EXCLUSIVE_CONFLICT",
          `exclusive group ${group.group_id} allows at most ${maxSelected}`,
        ),
      );
    }
  }
  const union = (field) =>
    [
      ...new Set(
        selectedRecords.flatMap((record) =>
          (record[field] ?? []).filter((value) => typeof value === "string"),
        ),
      ),
    ].sort();
  const componentMarks = [
    ...new Set(
      selectedRecords
        .map((record) => record.compatibility_mark)
        .filter((value) => typeof value === "string"),
    ),
  ].sort();
  return {
    composition: canonicalComposition,
    composition_hash:
      errors.length === 0
        ? objectHash(COMPOSITION_HASH_DOMAIN, canonicalComposition)
        : null,
    core_suite:
      corePaths.length === 1 && selectedRecords.length > 0
        ? clone(selectedRecords[0].core_suite)
        : null,
    required_suites: union("required_suites"),
    required_extensions: union("required_extensions"),
    required_crypto_profiles: union("required_crypto_profiles"),
    required_policy_categories: union("required_policy_categories"),
    component_compatibility_marks: componentMarks,
    errors,
  };
}
